README.md
# Ractive Window

jQueryUI Dialog-like windowing built on [Ractive](https://github.com/ractivejs/ractive)

## Where to get it?

Racive Window is available as a [giblet](https://github.com/evs-chris/gobble-giblet), a [component](https://github.com/componentjs/component), and a pre-assembled UMD module. Each flavor does not declare an explicit dependency on Ractive, but it is expected to be available as a global.

All of the pre-built files live in tags on the build branch.

### Development

Ractive Window uses [gobble](https://github.com/gobblejs/gobble) as its build tool, which makes it easy to build and play around with. The default mode in the gobble file is `development`, which will automatically pull in the edge version of Ractive and make it available along with the sandbox. There is an example file provided along with the source, which you can access by running gobble and pointing you browser at http://localhost:4567/sandbox/example.html.

## Usage

Ractive Window is a little bit weird as a Ractive component, because it's mostly built to be assembled dynamically rather than in pre-defined templates. What you get is a `Host` component, which manages a number of `Window` components. There is currently no way to manage windows in the template; all of the API is on the JS-side.

```html
<ractive-container/>
```

```js
// import RactiveWindow and Ractive using whatever module system first
Ractive.components.Host = RactiveWindow.Host;

var ractive = new Ractive({
  el: 'ractive-container',
  template: '<Host/>'
});

var host = ractive.findComponent('Host');
```

### Host

#### `host.newWindow(callback)`

`host.newWindow()` creates a new `Window` instance, assigns it a number, and returns a promise that resolves with the new instance. If a callback is supplied, it is called with instance just after the instance is created. The callback may return a promise, which is inserted into the init chain.

After the callback, the component's `activate` function is called (defaults to a no-op). If `activate` returns a promise, it is inserted into the init chain.

After `activate`, the window is moved to its displayed coordinates, which defaults to a sort of cascade positioning. If any coordinates were set in the init callback or `activate`, those will be used instead.

*TODO* `blockWindow`, `unblockWindow`, `raiseWindow`

### Window

Windows are composed of a number of partials that control appearance and content.

* title - the content of the title bar
* controls - the title bar control buttons
* body - the content of the window
* buttons - any dialog buttons
* foot - the status bar area

There are helpers to update each partial.

#### `window.content(tpl)`

This method resets the body partial of the window using the supplied template, which may be a string, pre-parsed template, or template function. The initial call to this method should probably be in the `newWindow()` init callback or window `activate` function.

#### `window.resize(width, height)`

This method resizes the window to the specified width and height. If the parameters are integers or strings with `px` units, then they will be used directly. If a unit other than `px` is used, each dimension is set to its given value and then reset to its corresponding pixel value, which is retrieved with `clientWidth` or `clientHeight`.

*TODO* `move()`, `title()`, `resizable()`, `minimize()`, `maximize()`, `restore()`, `raise()`, `kill()`, `buttons()`, `button()`, `close()`, `controls()`

### Window Events

Since events in Ractive components bubble up the hierarchy to the root, you can install handlers at the root for window events. You can use a wildcard event path or the component name, which will typically be `Window`. Be careful with wildcard events, especially with common names like `close`.
```js
// just components named Window
ractive.on('Window.close', (wnd) => doSomething());

// any component close event
ractive.on('*.close', (wnd) => doSomething());
```

* `close` - `listener(window)` - fires just before a window is torn down. If the window aborts closing from its handler, this will not be fired.
* `retitle` - `listener(title, window)` - fires when the window title changes. The first time it fires (first render), the title will be undefined.
* `maximize` - `listener(state, window)` - fires when the window is maximized.
* `minimize` - `listener(state, window)` - fires when the window is minimized.
* `restore` - `listener(state, window)` - fires when the window is restored.

States are defined as
```json
{
  "normal": 0,
  "minimized": 1,
  "maximized": 2
}
```

## License

Copyright (c) 2014 Chris Reeves. Released under an [MIT license](https://github.com/evs-chris/ractive-window/blob/master/LICENSE.md).


CHANGELOG.md
## 0.12.1

* Update default style.

## 0.12.0

* Adds support for template content in button labels and window titles.
* Adjusts handling of topmost windows.
* __BUG?:__ Keeps buttons on the bottom of maximized fixed windows.

## 0.11.0

* Adds support for having a single designated topmost window.
* Adds support for more easily colored titlebars.
* __BUG:__ Fixes resizing support in IE, which apparently likes *really* precise mouse coordinates.

## 0.10.2

* __BUG:__ Prevent text selection while dragging to resize or move windows.

## 0.10.1

* __BUG:__ Don't try to resize windows that aren't yet rendered using non-px sizes.

## 0.10.0

* Allow `resize` to be called with CSS units using dirty tricks.

## 0.9.1

* __BUG:__ Make sure windows that start out maximized restore to a visible position.

## 0.9.0

* Better default maximization styling using the newly added `maximized` window class.
* Internal events are now prefixed with `_` to make way for...
* There are now events for when a window title (retitle) changes, when the window state changes (maximize, minimize, restore), and when a window is closed (close).
* `WindowHost` may now have content, which it will just `{{yield}}`.

## 0.8.1

* __BUG?:__ `window.move` now returns a Promise.
* __BUG?:__ The Promise returned from `host.newWindow` will now resolve with the new window instead of `undefined`.

## 0.8.0

* Switched to gobble for bulding, and code is consequently ES6ified.
* Window creation process has been refactor to be slightly less wacky.

## 0.7.0

* Adds initial support for touch when moving and resizing windows.
* Adds support for button class at the window level.

## 0.6.0

* Changes the window render variable name to avoid clashes with more complex contents.
* Adds an afterClose promise that is resolved when a window is killed.
* Moves the example into its own directory and provides pre-built files.

## 0.5.0

* Window load callbacks that return promises are now chained with the promise returned from newWindow.
* Window and content div styles and classes can be added to the window model.

## 0.4.1

* __BUG:__ Fixes issue with leaky context and shared component state.

## 0.4.0

* Adds an element property to each window that references its actual DOM host.
* Adds a ractive window id.
* Changes template references to be relative to the current context.
* Makes cascade eventually wrap back to starting position instead of angling down-right forever.

## 0.3.0

* Adds initial support for cascade positioning.
* Fixes some minor css issues regarding fixed sized windows.
* Adds a few common attribute setters like title and resizable.

## 0.2.0

* Expands support for modal windows
* Adds support for dialog buttons
* Adds an initial message box helper
* Adds an initial relative position helper i.e. move('center')
* Adds support for fixed dialogs that size to fit their content


src/host.js
/* global Ractive */

import Window from "window";

var messageButtons = {
  ok: { label: 'OK', action: function() { this.result = 'ok'; this.close(); }, position: 'middle' },
  cancel: { label: 'Cancel', action: function() { this.result = 'cancel'; this.close(); }, position: 'middle' },
  yes: { label: 'Yes', action: function() { this.result = 'yes'; this.close(); }, position: 'middle' },
  no: { label: 'No', action: function() { this.result = 'no'; this.close(); }, position: 'middle' }
};

var WindowHost;
WindowHost = (function() {
  var counter = 0;

  // creates a window by appending to the slot array and retrieving the added window component
  // the callback is called first, if supplied
  // the window is then rendered and has its activate function called, if it has one
  // the window is then cascaded if it isn't already positioned
  function newWindow(e, cb) { // e can be an event or the callback. the callback (cb) is optional
    var current = counter;
    counter += 1;
    var host = this;
    return host.push('windowSlots', current).then(function() {
      var pr;

      // find the window instance let it know who it is
      var wnds = host.findAllComponents('Window');
      var wnd = wnds[wnds.length - 1];

      host.set('windows.' + current, wnd);
      wnd.parentNumber = current;
      wnd.set({
        'geometry.index': 1000 + wnds.length,
        'geometry.left': -9999,
        'geometry.top': -9999,
        'id': current
      });

      // call the creation callback if given
      var step1 = function() {
        var mpr;
        if (!!cb && typeof(cb) === 'function') {
          try {
            mpr = cb(wnd);
            if (!!mpr && typeof mpr.then === 'function') return mpr;
          } catch (e1) { console.log(e1); }
        } else if (typeof(e) === 'function') {
          try {
            mpr = e(wnd);
            if (!!mpr && typeof mpr.then === 'function') pr = mpr;
          } catch (e2) { console.log(e2); }
        }
      };
      pr = step1();

      // render the window and call its activate function
      var step2 = function() {
        var mpr;
        wnd.raise();
        return wnd.set('_wnd_rendered', true).then(function() {
          wnd.element = wnd.find('.ractive-window');
          try {
            mpr = wnd.activated();
            if (!!mpr && typeof mpr.then === 'function') return mpr;
          } catch (e4) { console.log(e4); }
        });
      };
      if (!!pr) pr = pr.then(step2); else pr = step2();

      // move the window from initial offscreen coordinates
      var step3 = function() {
        var mpr;
        if (wnd.get('geometry.left') === -9999) {
          return wnd.move('cascade').then(() => wnd);
        }
        return wnd;
      };
      if (!!pr) pr = pr.then(step3); else pr = step3();

      // return a promise for the last steps completion
      return pr;
    });
  }

  // shows a centered modal messagebox
  function messageBox(opts) {
    var args = arguments;
    var host = this;
    return new Ractive.Promise(function(y) {
      host.newWindow(function(w) {
        var message;
        if (args.length >= 2) {
          message = args[0];
          opts = args[1];
        } else if (args.length === 1 && typeof args[0] === 'string') {
          message = args[0];
          opts = {};
        }

        w.set('title', opts.title || 'Message');
        w.set('resizable', false);
        w.controls('close');
        w.content(message);
        var btns = opts.buttons || ['ok'], out = [];
        for (var i = 0; i < btns.length; i++) if (messageButtons.hasOwnProperty(btns[i])) out.push(messageButtons[btns[i]]);
        w.buttons(out);
        w.onClose = function() {
          this.kill();
          y(w.result || 'none');
        };
        if (!opts.hasOwnProperty('modal') || opts.modal) host.set('globalBlock', w);
        w.activated = function() { w.move('center'); };
      });
    });
  }

  return Ractive.extend({
    isolated: true,
    defaults: {
      control: {
        label: function label(control, lbl) { Window.partials[control + 'ControlLabel'] = lbl; }
      },
      controls: function() {
        var partial = '';
        for (var i = 0; i < arguments.length; i++) {
          partial += '{{>' + arguments[i] + 'Control}}';
        }
        Window.partials.controls = partial;
      }
    },
    components: { Window: Window },
    data: { windowSlots: [], windows: {}, blocks: {}, globalBlock: null },
    computed: { blocked: function() { return !!this.get('globalBlock'); } },
    template: "<div class='ractive-window-host-modal' style='{{^blocked}}display: none;{{/blocked}}'></div><div class='host-content'>{{yield}}</div>{{#windowSlots}}<Window/>{{/windowSlots}}",
    newWindow: newWindow,
    killWindow: function(wnd) {
      var blocks = this.get('blocks');
      var wnds = this.get('windows');
      var topWnd, topIdx = -1, i;
      if (!!wnds) {
        for (var w in wnds) {
          if (wnds[w] === wnd) delete wnds[w];
          else {
            i = wnds[w].get('geometry.index');
            if (i > topIdx) {
              topIdx = i;
              topWnd = wnds[w];
            }
          }
        }
        if (topWnd && !topWnd.get('topmost')) topWnd.set('topmost', true);
      }
      var slots = this.get('windowSlots');
      if (!!slots) {
        this.splice('windowSlots', slots.indexOf(wnd.parentNumber), 1);
      }
      for (i in blocks) {
        var arr = blocks[i];
        if (!!arr && Array.isArray(arr) && arr.indexOf(wnd.parentNumber) >= 0) arr.splice(arr.indexOf(wnd.parentNumber), 1);
      }
      if (wnd === this.get('globalBlock')) this.set('globalBlock', null);
      this.unblockUnblockedWindows();
    },
    raiseWindow: function(wnd) {
      var wndso = this.get('windows');
      var slots = this.get('windowSlots');
      var blocks = this.get('blocks');
      var wnds = [];
      var target = this.topLevelBlockers(wnd);
      target.push(wnd);

      for (var k in wndso) if (target.indexOf(wndso[k]) < 0) wnds.push(wndso[k]);

      // put windows in existing order in an array
      wnds.sort(function(a, b) {
        var ai = a.get('geometry.index'), bi = b.get('geometry.index');
        if (ai < bi) return -1; else if (ai > bi) return 1; else return 0;
      });

      // add the window being raised to the end-o-the-queue
      if (!!wnd) wnds = wnds.concat(target);

      // move blocked windows to below their blocker with splice and indexOf
      function moveBeforeBlocker(wnd, blockers) {
        for (var i in blockers) {
          var bl = wndso[blockers[i]];
          var wi = wnds.indexOf(wnd), bi = wnds.indexOf(bl);
          if (!!!bl || wi < 0 || bi < 0) continue;

          // blocker must be moved first in case it too is blocked, so recursion required
          var arr = blocks[bl.parentNumber];
          if (!!!arr && Array.isArray(arr) && arr.length > 0) moveBeforeBlocker(bl, arr);
          if (wi > bi) {
            wnds.splice(wi, 1); // remove window
            wnds.splice(bi, 0, wnd); // pop it in before bl
          }
        }
      }

      var i;
      for (i in slots) {
        var arr = blocks[slots[i]];
        if (!!arr && Array.isArray(arr) && arr.length > 0) moveBeforeBlocker(wndso[slots[i]], arr);
      }

      // loop through array assigning indices
      for (i in wnds) {
        wnds[i].set('geometry.index', 1000 + (+i));
        if (wnds[i] !== wnd) wnds[i].set('topmost', false);
      }
      if (!wnd.get('topmost')) wnd.set('topmost', true);

      function globalBlocks(wnd) {
        var res = [];
        if (!!!wnd) return res;

        var arr = blocks[wnd.parentNumber];
        if (!!arr && Array.isArray(arr) && arr.length > 0) {
          for (var i in arr) {
            res = res.concat(globalBlocks(wndso[arr[i]]));
          }
        }

        res.push(wnd);

        return res;
      }

      // if globalBlock is set, add 10000 to it and all of its blockers
      var globals = globalBlocks(this.get('globalBlock'));
      for (i in globals) globals[i].add('geometry.index', 10000);
    },
    topLevelBlockers: function(wnd) {
      if (!!!wnd) return [];

      var blocks = this.get('blocks');
      var wndso = this.get('windows');
      var arr = blocks[wnd.parentNumber];
      var res = [];

      if (!!!arr || !Array.isArray(arr) || arr.length === 0) return res;

      for (var i in arr) {
        var arr2 = blocks[arr[i]];
        if (!!!arr2 || !Array.isArray(arr2) || arr2.length === 0) res.push(wndso[arr[i]]);
        else {
          res = res.concat(this.topLevelBlockers(wndso[arr[i]]));
        }
      }

      return res;
    },
    blockWindow: function(target, blocker /*, ...*/) {
      if (!!!target || !!!blocker) return;

      var blocks = this.get('blocks');
      var arr = blocks[target.parentNumber];
      if (!!!arr || !Array.isArray(arr)) arr = [];
      if (arr.indexOf(blocker.parentNumber) < 0) arr.push(blocker.parentNumber);
      blocks[target.parentNumber] = arr;
      for (var i = 2; i < arguments.length; i++) {
        if (arr.indexOf(arguments[i].parentNumber) < 0) arr.push(arguments[i].parentNumber);
      }
      if (arr.length > 0) target.set('blocked', true);
      this.raiseWindow();
    },
    unblockWindow: function(target, blocker /*, ...*/) {
      if (!!!target | !!!blocker) return;

      var blocks = this.get('blocks');
      var arr = blocks[target.parentNumber];
      if (!!!arr || !Array.isArray(arr)) return;
      if (arr.indexOf(blocker.parentNumber) >= 0) arr.splice(arr.indexOf(blocker.parentNumber), 1);
      for (var i = 2; i < arguments.length; i++) {
        if (arr.indexOf(arguments[i].parentNumber) >= 0) arr.splice(arr.indexOf(arguments[i].parentNumber), 1);
      }
      if (arr.length === 0) target.set('blocked', false);
      this.raiseWindow();
    },
    unblockUnblockedWindows: function() {
      var blocks = this.get('blocks');
      var wndso = this.get('windows');
      for (var i in blocks) {
        var arr = blocks[i];
        if (!!!arr || !Array.isArray(arr) || arr.length === 0) {
          var wnd = wndso[i];
          if (!!wnd) wnd.set('blocked', false);
        }
      }
    },
    messageBox: messageBox
  });
})();

export default WindowHost;


gobblefile.js
var gobble = require('gobble');
var path = require('path');

var parts = [], dev = gobble.env() !== 'production';

var transpilerOptions = {
  blacklist: ['modules', 'useStrict']
};

// in dev mode, make libraries and the sandbox available
if (dev) {
  parts.push(gobble('sandbox').moveTo('sandbox'));
  parts.push(gobble('giblets').transform('giblets').moveTo('lib'));
}

var source = gobble('src');

var es5source = gobble(source).transform('6to5', transpilerOptions);

// build the main UMD module
var moduled = gobble(es5source)
  .transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'ractive-window.js',
    type: 'umd',
    name: 'RactiveWindow',
    strict: true
  });

// transpile from es6
var compiled = gobble(source)
  .transform('6to5', transpilerOptions);

// for production, build the component too
if (!dev) {
  parts.push(gobble(es5source).transform('esperanto-bundle', {
    entry: 'index.js',
    dest: 'component/index.js',
    type: 'cjs',
    strict: true
  }));
}

parts.push(moduled.
  transform('relative-sourcemaps', { prefix: 'ractive-window' })
);

// copy the assets
parts.push(gobble('assets').moveTo('assets'));

module.exports = gobble(parts);


src/window.js
/* global Ractive */

var template =
`{{#_wnd_rendered}}<div id='ractive-window-{{.id}}' class='ractive-window{{#(.buttons.length > 0)}} with-buttons{{/}}{{#.resizable}} resizable{{else}} fixed{{/}}{{#.geometry.state === 2}} maximized{{/}}{{#.class.window}} {{.class.window}}{{/}}{{#.topmost}} topmost{{/}}' on-click='_raise' style='{{#.hidden}}display: none;{{/}}top: {{.geometry.top}}px; left: {{.geometry.left}}px; {{#(.resizable || .geometry.state === 2)}}width: {{.geometry.width}}{{.geometry.dwunit}}; height: {{.geometry.height}}{{.geometry.dhunit}}; {{/}}z-index: {{.geometry.index}};{{#.style.window}} {{.style.window}}{{/}}'>
  <div class='rw-modal' on-mousedown='_moveStart' style='{{^.blocked}}display: none;{{/}}'></div>
  <div class='rw-interior'>
    <div class='rw-controls'>{{>controls}}</div>
    <div class='rw-title' on-touchstart-mousedown='_moveStart' on-dblclick='_restore'>{{>title}}</div>
    <div class='rw-body{{#.class.body}} {{.class.body}}{{/}}' {{#.style.body}}style='{{.style.body}}'{{/}}>{{>body}}</div>
    {{#(.buttons.length > 0)}}<div class='rw-buttons'>{{>buttons}}</div>{{/}}
    <div class='rw-resize-handle' on-touchstart-mousedown='_resizeStart'></div>
    <div class='rw-foot'>{{>foot}}</div>
  </div>
</div>{{/}}`;

var Window;
Window = Ractive.extend({
  template: template,
  onconstruct: function(opts) {
    var wnd = this;

    var sx, sy;
    var moveFn;
    moveFn = function(e) {
      var x, y;
      e.preventDefault();
      if (e.type.indexOf('touch') >= 0) {
        x = +(e.changedTouches[0].clientX);
        y = +(e.changedTouches[0].clientY);
      } else {
        x = +(e.x || e.clientX);
        y = +(e.y || e.clientY);
      }

      wnd.move(+wnd.get('geometry.left') + x - +sx, +wnd.get('geometry.top') + y - +sy);
      sx = x;
      sy = y;
      if (e.type === 'mouseup' || e.type === 'touchend') {
        document.removeEventListener('mousemove', moveFn, false);
        document.removeEventListener('mouseup', moveFn, false);
        document.removeEventListener('touchmove', moveFn, false);
        document.removeEventListener('touchend', moveFn, false);
      }
    };
    wnd.on('_moveStart', function(e) {
      if ((e.original.type === 'mousedown' && e.original.button === 0) || e.original.type === 'touchstart') {
        wnd.restore();
        if (e.original.type.indexOf('touch') >= 0) {
          sx = +(e.original.changedTouches[0].clientX);
          sy = +(e.original.changedTouches[0].clientY);
        } else {
          sx = +(e.original.x || e.original.clientX);
          sy = +(e.original.y || e.original.clientY);
        }
        document.addEventListener('mousemove', moveFn);
        document.addEventListener('mouseup', moveFn);
        document.addEventListener('touchmove', moveFn);
        document.addEventListener('touchend', moveFn);
        e.original.preventDefault();
      }
    });

    var resizeFn;
    resizeFn = function(e) {
      var x, y;
      e.preventDefault();
      if (e.type.indexOf('touch') >= 0) {
        x = e.changedTouches[0].clientX;
        y = e.changedTouches[0].clientY;
      } else {
        x = +(e.x || e.clientX);
        y = +(e.y || e.clientY);
      }
      var w = +wnd.get('geometry.width') + (x - +sx);
      var h = +wnd.get('geometry.height') + (y - +sy);
      wnd.resize(w, h);
      sx = x;
      sy = y;
      if (e.type === 'mouseup' || e.type === 'touchend') {
        document.removeEventListener('mousemove', resizeFn, false);
        document.removeEventListener('mouseup', resizeFn, false);
        document.removeEventListener('touchmove', resizeFn, false);
        document.removeEventListener('touchend', resizeFn, false);
      }
    };
    wnd.on('_resizeStart', function(e) {
      if ((e.original.type == 'mousedown' && e.original.button === 0) || e.original.type === 'touchstart') {
        wnd.restore();
        if (e.original.type.indexOf('touch') >= 0) {
          sx = e.original.changedTouches[0].clientX;
          sy = e.original.changedTouches[0].clientY;
        } else {
          sx = (e.original.x || e.original.clientX);
          sy = (e.original.y || e.original.clientY);
        }
        document.addEventListener('mousemove', resizeFn);
        document.addEventListener('mouseup', resizeFn);
        document.addEventListener('touchmove', resizeFn);
        document.addEventListener('touchend', resizeFn);
      }
    });

    var stateFn = function(target, e) {
      switch (target) {
        case 'min': wnd.minimize(); break;
        case 'max': wnd.maximize(); break;
        case 'normal': wnd.restore(); break;
        default: break;
      }
    };

    wnd.on('_minimize', function(e) { stateFn('min', e); });
    wnd.on('_restore', function(e) {
      switch (wnd.get('geometry.state')) {
        case 0: stateFn('max', e); break;
        case 1:
          case 2:
          stateFn('normal', e); break;
        default: break;
      }
    });
    wnd.on('_raise', function(e) { wnd.raise(); });
    wnd.on('_close', function(e) { wnd.close(); });
    wnd.on('_dialog-button', function(e) {
      var fn = e.context.action;
      if (!!fn && typeof fn === 'function') fn.call(this);
    });

    wnd.result = null;
    wnd.waitForClose = wnd.afterClose = new Ractive.Promise(function(y, n) {
      var fn = function(t) {
        return function(v) {
          wnd.completeAfterClose = null;
          wnd.rejectAfterClose = null;
          t(v);
        };
      };
      wnd.completeAfterClose = fn(y);
      wnd.rejectAfterClose = fn(n);
    });
  },
  onrender() {
    if (!!!this.get('buttonClass') && !!this.parent.get('buttonClass')) {
      this.set('buttonClass', this.parent.get('buttonClass'));
    }

    this.watchers = this.observe({
      title: (n, o) => {
        this.fire('retitle', n, this);
      },

      'geometry.state': (n, o) => {
        switch (n) {
          case 0: this.fire('restore', n, this); break;
          case 1: this.fire('minimize', n, this); break;
          case 2: this.fire('maximize', n, this); break;
        }
      }
    });
  },
  onunrender() { if (this.watchers && typeof this.watchers.cancel === 'function') this.watchers.cancel(); },
  activated() {},
  data() { return {
    _wnd_rendered: false,
    blocked: false,
    resizable: true,
    geometry: {
      top: -9999, left: -9999, width: 200, height: 200, state: 0, dwunit: 'px', dhunit: 'px', index: 1000,
      minimum: { x: 0, y: 0, width: 70, height: 50 }
    },
    style: {},
    class: {},
    makePartial(key, template) {
      if (!this._makePartial_templates) this._makePartial_templates = {};
      if (this._makePartial_templates[key] != template) {
        this.resetPartial(key, template);
        this._makePartial_templates[key] = template;
      }
      return key;
    }
  }; },
  partials: {
    title: "{{> ~/makePartial('titleTpl', .title) }}",
    body: '',
    foot: '',
    buttons: "{{#.buttons:i}}<button on-click='_dialog-button' class='{{.position || ''}}{{#.buttonClass}} {{.buttonClass}}{{/}}{{#../../class.button}} {{../../class.button}}{{/}}' disabled='{{!.enabled}}'>{{> ~/makePartial('button' + i + 'Tpl', .label) }}</button>{{/}}",
    controls: '{{>minimizeControl}}{{>restoreControl}}{{>closeControl}}',
    minimizeControl: "<button on-click='_minimize' class='rw-minimize'>{{>minimizeControlLabel}}</button>",
    minimizeControlLabel: "_",
    restoreControl: "<button on-click='_restore' class='rw-restore'>{{>restoreControlLabel}}</button>",
    restoreControlLabel: "^",
    closeControl: "<button on-click='_close' class='rw-close'>{{>closeControlLabel}}</button>",
    closeControlLabel: "X"
  },
  rerender: function() {
    var wnd = this;
    if (!wnd.get('_wnd_rendered')) return Ractive.Promise.resolve('ok');
    wnd.set('_wnd_rendered', false);
    return this.set('_wnd_rendered', true);
  },
  title: function(str) { this.set('title', str); },
  move: function(x, y) {
    if (typeof x === 'string') {
      switch (x) {
        case 'center':
          case 'centerScreen':
          return this.set({
          'geometry.top': (this.parent.el.clientHeight - this.element.clientHeight) / 2,
          'geometry.left': (this.parent.el.clientWidth - this.element.clientWidth) / 2
        });
        case 'cascade':
          return this.set({
          'geometry.top': ((this.parentNumber % 10) * 20) + 10,
          'geometry.left': ((this.parentNumber % 50) * 20) + 10
        });
      }
      return Ractive.Promise.resolve(false);
    }
    y = +y;
    x = +x;
    var min = this.get('geometry.minimum');
    var max = this.get('geometry.maximum');
    var w = +this.get('geometry.width');
    var h = +this.get('geometry.height');
    if (!!max) {
      if (x + w > +max.x) x = +max.x - x;
      if (y + h > +max.y) y = +max.y - y;
    }
    if (!!min) {
      if (x < +min.x) x = +min.x;
      if (y < +min.y) y = +min.y;
    }
    return this.set({
      'geometry.top': y,
      'geometry.left': x
    });
  },
  resize: function(w, h) {
    w = getDimPx.call(this,'width', w);
    h = getDimPx.call(this, 'height', h);
    var min = this.get('geometry.minimum');
    var max = this.get('geometry.maximum');
    if (!!max) {
      if (w > max.width) w = max.width;
      if (w > max.height) w = max.height;
    }
    if (!!min) {
      if (w < min.width) w = min.width;
      if (h < min.height) h = min.height;
    }
    this.set({
      'geometry.width': w,
      'geometry.height': h
    });
  },
  resizable: function(b) { this.set('resizable', b); },
  minimize: function() {
    var wnd = this;
    if (wnd.get('geometry.state') !== 1) {
      wnd.set({
        hidden: true,
        'geometry.state': 1
      });
      wnd.fire('minimized', { window: wnd });
    }
  },
  maximize: function() {
    var wnd = this;
    if (wnd.get('geometry.state') !== 2) {
      wnd.normalGeometry = { top: wnd.get('geometry.top'), left: wnd.get('geometry.left'), width: wnd.get('geometry.width'), height: wnd.get('geometry.height') };
      wnd.set({
        'hidden': false,
        'geometry.left': 0,
        'geometry.top': 0,
        'geometry.width': 100,
        'geometry.height': 100,
        'geometry.dwunit': '%',
        'geometry.dhunit': '%',
        'geometry.state': 2
      });
      wnd.fire('maximized', { window: wnd });
    }
  },
  restore: function() {
    var wnd = this;
    switch (wnd.get('geometry.state')) {
      case 1: // minimized
        wnd.set({
          hidden: false,
          'geometry.state': 0
        });
        break;
      case 2:
        var g = wnd.normalGeometry || {};
        wnd.normalGeometry = null;
        if (g.top < 0 || g.left < 0) {
          g.top = 0;
          g.left = 0;
        }
        wnd.set({
          hidden: false,
          'geometry.left': g.left,
          'geometry.top': g.top,
          'geometry.width': g.width,
          'geometry.height': g.height,
          'geometry.dwunit': 'px',
          'geometry.dhunit': 'px',
          'geometry.state': 0
        });
        break;
      default: break;
    }
    this.raise();
  },
  raise: function() {
    if (!!this.parent) this.parent.raiseWindow(this);
  },
  kill: function() {
    var wnd = this;
    this.fire('close', this);
    if (!!wnd.parent) {
      wnd.parent.killWindow(wnd);
    } else {
      wnd.teardown();
    }
    if (!!wnd.completeAfterClose) wnd.completeAfterClose(wnd.result);
  },
  content: function(ct) {
    return this.resetPartial('body', ct);
  },
  buttons: function() {
    var arr = [], i;
    this.set('buttons', arr);
    if (arguments.length === 1 && typeof arguments[0].length === 'number') {
      arr = arguments[0];
    } else {
      for (i = 0; i < arguments.length; i++) {
        arr.push(arguments[i]);
      }
    }
    var left = [], right = [], middle = [];
    for (i = 0; i < arr.length; i++) {
      var b = arr[i];
      if (!!b.position) {
        if (b.position === 'left') left.push(b);
        else if (b.position === 'right') right.push(b);
        else if (b.position === 'middle') middle.push(b);
        else if (b.position === 'center') middle.push(b);
        else { right.push(b); b.position = 'right'; }
      } else { right.push(b); b.position = 'right'; }
      if (!b.hasOwnProperty('enabled')) b.enabled = true;
    }
    arr = [];
    for (i = 0; i < left.length; i++) arr.push(left[i]);
    for (i = right.length - 1; i >= 0; i--) arr.push(right[i]);
    for (i = 0; i < middle.length; i++) arr.push(middle[i]);
    this.set('buttons', arr);
  },
  button: function(name, cb) {
    var arr = this.get('buttons');
    var btn, i;
    if (typeof name === 'number') { btn = arr[name]; i = name; }
    else for (i = 0; i < arr.length; i++) {
      if (arr[i].label === name) {
        btn = arr[i];
        break;
      }
    }

    if (!!btn) {
      cb(btn);
      this.set('buttons.' + i, btn);
    }
  },
  controls: function() {
    var arr = [], i, str = '';
    if (arguments.length === 1 && typeof arguments[0] !== 'string') arr = arguments[0];
    else {
      for (i = 0; i < arguments.length; i++) arr.push(arguments[i]);
    }
    for (i = 0; i < arr.length; i++) str += '{{>' + arr[i] + 'Control}}';
    this.partials.controls = str;
    return this.rerender();
  },
  onClose: function() {
    this.kill();
  },
  close: function(fn) {
    if (!!!fn) fn = this.onClose;
    if (fn.length === 0) fn.call(this);
    else {
      var wnd = this;
      fn.call(this, function(close) { if (close) wnd.kill(); });
    }
  }
});

var cssUnit = /([\d\.]+)(.*)/;
function getDimPx(dim, length) {
  var [ whole, size, unit ] = cssUnit.exec(length.toString());
  unit = unit || 'px';
  var dunit = dim === 'width' ? 'dwunit' : 'dhunit';
  var div = this.find('div');

  if (unit === 'px') {
    return size;
  } else if (div) {
    var toSet = {};
    toSet['geometry.' + dim] = size;
    toSet['geometry.' + dunit] = unit;
    this.set(toSet);
    var v = this.find('div')['client' + dim[0].toUpperCase() + dim.substring(1)];
    toSet['geometry.' + dim] = v;
    toSet['geometry.' + dunit] = 'px';
    this.set(toSet);
    return v;
  }
}

export default Window;
